packages/core/http/core-http-server-internal/src/http_service.ts

```diff
@@ -222,12 +225,65 @@ export class HttpService
         await this.httpsRedirectServer.start(config);
       }
 
+      if (config.oas.enabled) {
+        this.log.info('Registering experimental OAS API');
+        this.registerOasApi(config);
+      }
+
       await this.httpServer.start();
     }
 
     return this.getStartContract();
   }
 
+  private registerOasApi(config: HttpConfig) {
+    const basePath = this.internalSetup?.basePath;
+    const server = this.internalSetup?.server;
+    if (!basePath || !server) {
+      throw new Error('Cannot register OAS API before server setup is complete');
+    }
+
+    const baseUrl =
+      basePath.publicBaseUrl ?? `http://localhost:${config.port}${basePath.serverBasePath}`;
+
+    server.route({
+      path: '/api/oas',
+      method: 'GET',
+      handler: async (req, h) => {
+        const pathStartsWith = req.query?.pathStartsWith;
+        const pluginId = req.query?.pluginId;
+        return await firstValueFrom(
+          of(1).pipe(
+            HttpService.generateOasSemaphore.acquire(),
+            mergeMap(async () => {
+              try {
+                // Potentially quite expensive
```

Comment:
Kibana's start up time will increase depending on how expensive this route registration is. How big is this risk? If potentially ends up as 'almost always', I'd add it to the risks matrix in the PR description.

Reply:
I'll add a risks section as well :). The expensive operation is inside of the route handler so will not impact Kibana startup time. I defensively added a semaphore to allow only one of these operations to run on a given Kibana.